src/main/scala/com/crobox/clickhouse/stream/ClickhouseIndexingSubscriber.scala

```diff
@@ -279,35 +281,49 @@ class ClickhouseBulkActor(targetTable: String,
     buffer.clear()
     context.stop(self)
   }
-
-  val insertQuery = s"INSERT INTO $targetTable FORMAT JSONEachRow"
-
   private val objectMapper = new ObjectMapper()
     .registerModule(new JodaModule)
     .registerModule(new DefaultScalaModule)
     .setSerializationInclusion(JsonInclude.Include.NON_DEFAULT)
 
   private def index(): Unit = {
-    val payload = buffer.map(i => objectMapper.writeValueAsString(i.data)).mkString("\n") + "\n"
-    val count   = buffer.size
+    val payload = buffer
+      .collect {
+        case Insert(table, data) => data
+      }
+
+    val count = buffer.size
     logger.debug(s"Inserting $count")
+
     sent = sent + count
     logger.debug(s"Clickhouse $targetTable sent: $sent (confirmed: $confirmed, failed: $failed)")
-    send(payload, count)
+    send(targetTable, payload, count)
     buffer.clear()
 
     // buffer is now empty so no point keeping a scheduled flush after operation
     flushAfterScheduler.foreach(_.cancel)
     flushAfterScheduler = None
   }
 
-  private def send(payload: String, count: Int, retries: Int = 3): Unit =
-    client.execute(insertQuery, payload) onComplete {
-      case Failure(_) if retries > 0 => send(payload, count, retries - 1)
-      case Failure(e)                => self ! ClickhouseBulkActor.FlushFailure(e, count)
-      case Success(resp: String)     => self ! ClickhouseBulkActor.FlushSuccess(resp, count)
-    }
+  private def send(table: String, payload: Seq[Map[String, String]], count: Int): Unit = {
+    if (payload.isEmpty) Unit
+
+    val colList = payload.head.keys.mkString(", ")
```

Comment:
What about the scenario when you insert multiple rows with different columns. f.e.
```scala
Seq(Insert("test", Map("a" -> "b")), Insert("test", Map("b" -> "c")))
```